### PR TITLE
Group communicator on (parallel) cut meshes [cut-mesh-groups-dev]

### PR DESCRIPTION
**Pull request by rw-anderson**
This change facilitates the use of the group communicator mechanism on nonconforming meshes.  Slaved nodes are never shared, and therefore do not participate in group communications.  Slaved node values will still need to be modified by a conforming prolongation operator in order to get a conforming grid function.

TODO:
- [ ] Update `CHANGELOG`

### CODE REVIEW

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -4415,8 +4415,8 @@ void Mesh::GenerateNCFaceInfo()
       slave_fi.NCFace = nc_faces_info.Size();
       nc_faces_info.Append(NCFaceInfo(true, slave.master, &slave.point_matrix));
 
-      slave_fi.Elem2No = master_fi.Elem1No;
-      slave_fi.Elem2Inf = 64 * master_nc.MasterFace; // get lf no. stored above
+      // slave_fi.Elem2No = master_fi.Elem1No;
```
Comment (tzanio) [inline mesh/mesh.cpp]: Why are these not needed?

**Comment by tzanio** [comment]
Approved modulus todo items/questions.

File: mesh/pmesh.cpp [hunk 2]
```diff
@@ -87,605 +87,197 @@ ParMesh::ParMesh(MPI_Comm comm, Mesh &mesh, int *partitioning_,
...
-   // the first group is the local one
-   group.Recreate(1, &MyRank);
-   groups.Insert(group);
 
 #ifdef MFEM_DEBUG
    if (Dim < 3 && mesh.GetNFaces() != 0)
    {
-      mfem::err << "ParMesh::ParMesh (proc " << MyRank << ") : "
-                "(Dim < 3 && mesh.GetNFaces() != 0) is true!" << endl;
+      cerr << "ParMesh::ParMesh (proc " << MyRank << ") : "
```
Comment (jakubcerveny) [inline mesh/pmesh.cpp]: Is this incorrectly merged?

File: mesh/mesh.cpp (hunk 1 above)
Comment (jakubcerveny) [reply mesh/mesh.cpp]: I'm also reluctant to remove these.

**Comment by jakubcerveny** [comment]
I was trying to see if this functionality could fix AMR mesh visualization (it got broken by pmatrix-dev). In theory we should be able comment out the line `if (!pncmesh)` in `ParMesh::Print()` and its `else` branch and rely again on `sface_lface` when printing shared faces. However, I get strange results:
glvis_s01
Why is that?

**Comment by jakubcerveny** [comment]
Another question: from the diffs it seems that the initialization of shared object tables is done in the ParMesh constructor. What happens if the mesh gets refined? Are the tables updated?

**Comment by rw-anderson** [comment]
Jakub, what is that 3D image from?  Is that glvis?  Or VisIt?  What type of finite element space are you using?

**Comment by jakubcerveny** [comment]
Hi Bob, the image is from ex1p on two processors (just a simple H1 space), visualized with GLVis. It is before slave-face-fix was merged and I was trying to see if the table 'sface_lface' works for printing slave faces in NC meshes. It seems really suspicious that an internal face was printed.

**Comment by rw-anderson** [comment]
Thanks, I will take a look.

**Comment by rw-anderson** [comment]
Jakub, 

The partition I get on a mesh like the one you show on two processors looks like this:

2-hex-amr-init

When I print the shared faces, I get 5 of them, which seems like the correct number.  The shared entities are the ones which are shared on the cut mesh.  The slave faces are not treated as shared.  Does that make sense?

**Comment by jakubcerveny** [comment]
Hi Bob, the mesh from my screenshot was obtained programmatically directly in ex1p. After the mesh (consisting of 2 hexes) is distributed, I refine element 0 on rank 0 (rank 1 needs to be passed an empty refinement array). I don't rebalance. Also, this was before `slave-faces-fix` (PR #503), which may now obscure the problem. Modifying `ParMesh::Print` as described above should reveal it again.

**Comment by rw-anderson** [comment]
Ok, I refined then partitioned.  Let me try it the other way.

**Comment by rw-anderson** [comment]
I should also mention that I tried to specify the partitioning by passing in a partitioning array into the ParMesh constructor, and that seemed to cause problems.  Do you expect that should work on an NC mesh?  The ParNCMesh seems to specify its own equipartition independently of what was passed in.

**Comment by rw-anderson** [comment]
I see what's going on in this case.  This relates to your other comment about the groups being set up in the constructor.  These data structures need to be recomputed on refinement.  I'm working on that.

(Specifically, the partition still says that element 0 is on 0 and element 1 is on 1 after refinement, giving the shared face shown.)

**Comment by rw-anderson** [comment]
The dynamic updates of the communication groups on the cut mesh has been deferred for the next merge.

**Comment by camierjs** [comment]
I'm getting Segmentation fault in Laghos master since this merge.
I'm looking into it.

**Comment by v-dobrev** [comment]
This PR was reverted in the master branch because it was causing issues in Laghos and in some of the internal LLNL tests.

It should be possible to recreate (or simply reopen) this PR, and resolve the issues before merging into the master branch.

**Pull request merged**